### ISSUE TITLE
flatiter attributes produce wrong values

### ISSUE THREAD

**Issue by HagaiHargil** [numpy/numpy#17042]
When creating a `flatiter` object, the values are returned correctly, but the attributes are wrong.

### Reproducing code example:
```python
import numpy as np

arr = np.arange(9).reshape((3, 3))
fl = arr.flat
for item in fl:
    print(item)  # correct
    print(fl.index)  # wrong, starts from 1
    print(fl.coords)  # wrong, offset by 1 and will end at (3, 0) instead of (2, 2)
 ```

### Numpy/Python version information:

1.19.1 3.7.7 (default, May  7 2020, 21:25:33)
[GCC 7.3.0]
[Tested on 3.8 as well with the same results]

**Comment by mattip** [comment]
When you access `fl` in the loop, you have incremented `index`. Why is that wrong? The example in [the documentation](https://numpy.org/doc/stable/reference/generated/numpy.flatiter.index.html#numpy.flatiter.index) shows how to check it before you enter the loop if you want `0`. The [coords](https://numpy.org/doc/stable/reference/generated/numpy.flatiter.coords.html#numpy.flatiter.coords) is documented as "the current coordinates", so the "last" one (after 9 calls to `next(fl)`) is undefined since obtaining the current value via `next(fl)` will raise an error.

**Comment by HagaiHargil** [comment]
Thanks for the quick reply, but there are two issues with your proposed solution:

1. The `fl` iterator has the correct value before the loop starts, which is great. The issue is that the value (the one you access with `fl[index]`) remains constant while the `coords` and `index` attributes do move on to the next element in the sequence. The "interaction" between the `flatiter` object and the for loops seems off.

2. The last coordinate that `fl.coords` produces is `(3, 0)` in this case, which is very wrong regardless of your thoughts of my previous comment :)

**Comment by mattip** [comment]
1. Depends how you think about what "current" means in the context of an iterator. Your mental model is different than NumPy's. Perhaps you should give a little more motivation for your concept: what use case do you have for accessing the `index` and/or the `coords`? You could propose changing these to support your use case, but it must be well thought out.

2. I claim accessing any attribute of an expired iterator is not defined.

Perhaps an `nditer` would suit your needs better, as it provides a  way to [track the indices](https://numpy.org/devdocs/reference/arrays.nditer.html?highlight=iterating#tracking-an-index-or-multi-index)

**Comment by eric-wieser** [comment]
I think I agree with @HagaiHargil that the current behavior is bad. My best guess is that it was inherited from `nditer`, which was originally designed to be used as:
```
while not it.done:
    foo(it.value)
    it.iternext()`
```
where the python iterator protocol was seemingly added as an afterthought.

Perhaps the answer is that `flatiter` is a very thin wrapper around the C api, and this style of `while` loop is how the resulting C code appears.

I think we should change this _if we think we can get away with breaking everyone currently using it_.

**Comment by seberg** [comment]
Well, we will have to have a transition, otherwise we assume nobody uses this.

The only way to achieve that would seem to me:

1. We add new attributes, the old ones can be deprecated. Requires a new name though?
2. We assume nobody much uses it, so just go through a noisy and unavoidable FutureWarning. (I do assume the number of users is very small, its a fringe feature and if it was large this should have been reported years ago. I guess the way to avoid the warning )
3. We create new API of the form `for index, elem in arr.iter(axis=None, index=True):`, maybe it could be tagged on to the `arr.flat` name.

I guess `nditer` can be used to achieve this already, but frankly, its too heavy machinery for most users.  @eric-wieser I doubt its inherited from `nditer`, I think `nditer` is much newer than the flat-iter (and other iteration C-API). But it definitely is inherited from the same C looping pattern...

Btw. is there much of a precedence for this type of iteration in Python or large packages?
```
for a in iter:
    something = iter.attribute
```
it looks to me like something strange, especially since it requires the `flat = arr.flat` use before the loop (or walrus operator).

**Comment by eric-wieser** [comment]
> Btw. is there much of a precedence for this type of iteration in Python or large packages?

I think that's a fair point - I can't think of any precedent, although I'd be pleased to find out if any exists. Perhaps we should just deprecate this usage in favor of a better API, as you suggest.

**Comment by eric-wieser** [comment]
Furthermore, I can think of an argument against that API design - it breaks pretty badly for things like `heapq.merge`, or `itertools.tee`, where the caching of a future item would update the indices to be wrong.

**Comment by seberg** [comment]
@HagaiHargil actually, we do have `np.ndenumerate`, I suggest you use that instead.

**Comment by HagaiHargil** [comment]
Thanks for your responses.

@mattip My use-case involves iterating over a view of some array while (a) mutating it and (b) mutating another array at approximately the same coordinates. To do that I want a flattened version of the array as well as the "underlying" coordinates of this view.

As for my two cents on this API - I also think it's not very Pythonic and weird in general, but I do believe that the bug that makes it possible for the `coords` attribute to return a none-existing coordinate should probably be fixed regardless of the final decision on the new API. I disagree with @mattip with his observation that the iterator is "expired" and thus can cause _undefined behavior_ of some sort, since nothing signals to the user that this iterator has expired. There's no `StopIteration`, for example.

It didn't occur to me that I can use the walrus operator here, so I'm kinda happy that @seberg pointed that out :) I haven't had the chance to use it in any real code before...

**Comment by eric-wieser** [comment]
> @HagaiHargil actually, we do have `np.ndenumerate`, I suggest you use that instead.

Perhaps we should deprecate attribute access on `.flat` with a warning suggesting to use `ndenumerate`?

**Comment by seberg** [comment]
Yes, I agree... We have decent alternatives, and it seem a pretty buggy thing to use. It was probably just exposed to Python without much thought, and only makes sense in the C-side API iteration pattern. Which means that @HagaiHargil should simply not use this. `np.ndenumerate` doesn't provide a linear index, but I wonder if that matters much.
